File: csharp/Wjybxx.Dson.Core/src/DsonInt32.cs
#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Wjybxx.Dson
{
/// <summary>
/// DsonInt32
/// </summary>
public sealed class DsonInt32 : DsonNumber, IEquatable<DsonInt32>, IComparable<DsonInt32>, IComparable
{
    private readonly int _value;

    public DsonInt32(int value) {
        this._value = value;
    }

    public override DsonType DsonType => DsonType.Int32;
    public int Value => _value;

    public override int IntValue => _value;
    public override long LongValue => _value;
    public override float FloatValue => _value;
    public override double DoubleValue => _value;

    #region equals

    public bool Equals(DsonInt32? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _value == other._value;
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((DsonInt32)obj);
    }

    public override int GetHashCode() {
        return _value;
    }

    public static bool operator ==(DsonInt32? left, DsonInt32? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonInt32? left, DsonInt32? right) {
        return !Equals(left, right);
    }

    public int CompareTo(DsonInt32? other) {
        if (ReferenceEquals(this, other)) return 0;
        if (ReferenceEquals(null, other)) return 1;
        return _value.CompareTo(other._value);
    }

    public int CompareTo(object? obj) {
        if (ReferenceEquals(null, obj)) return 1;
        if (ReferenceEquals(this, obj)) return 0;
        return obj is DsonInt32 other ? CompareTo(other) : throw new ArgumentException($"Object must be of type {nameof(DsonInt32)}");
    }

    public static bool operator <(DsonInt32? left, DsonInt32? right) {
        return Comparer<DsonInt32>.Default.Compare(left, right) < 0;
    }

    public static bool operator >(DsonInt32? left, DsonInt32? right) {
        return Comparer<DsonInt32>.Default.Compare(left, right) > 0;
    }

    public static bool operator <=(DsonInt32? left, DsonInt32? right) {
        return Comparer<DsonInt32>.Default.Compare(left, right) <= 0;
    }

    public static bool operator >=(DsonInt32? left, DsonInt32? right) {
        return Comparer<DsonInt32>.Default.Compare(left, right) >= 0;
    }

    #endregion

    public override string ToString() {
        return $"{nameof(DsonType)}: {DsonType}, {nameof(_value)}: {_value}";
    }

    #region 池化管理

    internal const int POOL_START = -9;
    internal const int POOL_END = 127;
    // 注意初始化顺序
    private static readonly DsonInt32[] POOL = new DsonInt32[POOL_END - POOL_START + 1];
    public static readonly DsonInt32 ZERO;
    public static readonly DsonInt32 ONE;
    public static readonly DsonInt32 MINUS_ONE;

    static DsonInt32() {
        for (int i = POOL_START; i <= POOL_END; i++) {
            POOL[i - POOL_START] = new DsonInt32(i);
        }
        ZERO = ValueOf(0);
        ONE = ValueOf(1);
        MINUS_ONE = ValueOf(-1);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static DsonInt32 ValueOf(int value) {
        if (value < POOL_START || value > POOL_END) {
            return new DsonInt32(value);
        }
        return POOL[value - POOL_START];
    }

    #endregion
}
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonInt32.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import javax.annotation.Nonnull;

/**
 * @author wjybxx
 * date - 2023/4/19
 */
public final class DsonInt32 extends DsonNumber implements Comparable<DsonInt32> {

    private final int value;

    public DsonInt32(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.INT32;
    }

    @Override
    public Integer number() {
        return value;
    }

    @Override
    public int intValue() {
        return value;
    }

    @Override
    public long longValue() {
        return value;
    }

    @Override
    public float floatValue() {
        return value;
    }

    @Override
    public double doubleValue() {
        return value;
    }

    //region equals

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DsonInt32 dsonInt32 = (DsonInt32) o;

        return value == dsonInt32.value;
    }

    @Override
    public int hashCode() {
        return value;
    }

    @Override
    public int compareTo(DsonInt32 that) {
        return Integer.compare(value, that.value);
    }

    // endregion

    @Override
    public String toString() {
        return "DsonInt32{" +
                "value=" + value +
                '}';
    }

    // region 池化管理

    static final int POOL_START = -9;
    static final int POOL_END = 127;
    // 注意初始化顺序
    private static final DsonInt32[] POOL = new DsonInt32[POOL_END - POOL_START + 1];
    public static final DsonInt32 ZERO = valueOf(0);
    public static final DsonInt32 ONE = valueOf(1);
    public static final DsonInt32 MINUS_ONE = valueOf(-1);

    static {
        for (int i = POOL_START; i <= POOL_END; i++) {
            POOL[i - POOL_START] = new DsonInt32(i);
        }
    }

    public static DsonInt32 valueOf(int value) {
        if (value < POOL_START || value > POOL_END) {
            return new DsonInt32(value);
        }
        return POOL[value - POOL_START];
    }
    // endregion
}